RecoTauTag/RecoTau/plugins/PATTauDiscriminationByMVAIsolationRun2.cc

```diff
@@ -0,0 +1,306 @@
+  if ( tau->leadChargedHadrCand().isNull() ) return 0.;
+	
+  int tauDecayMode = tau->decayMode();
+	
+  if ( ((mvaOpt_ == kOldDMwoLT || mvaOpt_ == kOldDMwLT || mvaOpt_ == kDBoldDMwLT || mvaOpt_ == kPWoldDMwLT) && (tauDecayMode == 0 || tauDecayMode == 1 || tauDecayMode == 2 || tauDecayMode == 10)) ||
+     ((mvaOpt_ == kNewDMwoLT || mvaOpt_ == kNewDMwLT || mvaOpt_ == kDBnewDMwLT || mvaOpt_ == kPWnewDMwLT) && (tauDecayMode == 0 || tauDecayMode == 1 || tauDecayMode == 2 || tauDecayMode == 5 || tauDecayMode == 6 || tauDecayMode == 10)) ) {
+	
+    float chargedIsoPtSum = tau->tauID(chargedIsoPtSums_);
+    float neutralIsoPtSum = tau->tauID(neutralIsoPtSums_);
+    float puCorrPtSum     = tau->tauID(puCorrPtSums_);
+    float photonPtSumOutsideSignalCone = tau->tauID(photonPtSumOutsideSignalCone_);
+    float footprintCorrection = tau->tauID(footprintCorrection_);
+		
+    float decayDistX = tau->flightLength().x();
+    float decayDistY = tau->flightLength().y();
+    float decayDistZ = tau->flightLength().z();
+    float decayDistMag = std::sqrt(decayDistX*decayDistX + decayDistY*decayDistY + decayDistZ*decayDistZ);
+		
+    // --- The following 5 variables differ slightly between AOD & MiniAOD
+    //     because they are recomputed using packedCandidates saved in the tau
+    float nPhoton = (float)clusterVariables_.tau_n_photons_total(*tau);
+    float ptWeightedDetaStrip = clusterVariables_.tau_pt_weighted_deta_strip(*tau, tauDecayMode);
+    float ptWeightedDphiStrip = clusterVariables_.tau_pt_weighted_dphi_strip(*tau, tauDecayMode);
+    float ptWeightedDrSignal = clusterVariables_.tau_pt_weighted_dr_signal(*tau, tauDecayMode);
+    float ptWeightedDrIsolation = clusterVariables_.tau_pt_weighted_dr_iso(*tau, tauDecayMode);
+    // ---
+    float leadingTrackChi2 = tau->leadingTrackNormChi2();
+    float eRatio = clusterVariables_.tau_Eratio(*tau);
+		
+    if ( mvaOpt_ == kOldDMwoLT || mvaOpt_ == kNewDMwoLT ) {
+      mvaInput_[0]  = std::log(std::max((float)1., (float)tau->pt()));
+      mvaInput_[1]  = std::fabs((float)tau->eta());
+      mvaInput_[2]  = std::log(std::max((float)1.e-2, chargedIsoPtSum));
+      mvaInput_[3]  = std::log(std::max((float)1.e-2, neutralIsoPtSum - (float)0.125*puCorrPtSum));
+      mvaInput_[4]  = std::log(std::max((float)1.e-2, puCorrPtSum));
+      mvaInput_[5]  = tauDecayMode;
+    } else if ( mvaOpt_ == kOldDMwLT || mvaOpt_ == kNewDMwLT  ) {
+      mvaInput_[0]  = std::log(std::max((float)1., (float)tau->pt()));
+      mvaInput_[1]  = std::fabs((float)tau->eta());
+      mvaInput_[2]  = std::log(std::max((float)1.e-2, chargedIsoPtSum));
+      mvaInput_[3]  = std::log(std::max((float)1.e-2, neutralIsoPtSum - (float)0.125*puCorrPtSum));
+      mvaInput_[4]  = std::log(std::max((float)1.e-2, puCorrPtSum));
+      mvaInput_[5]  = tauDecayMode;
+      mvaInput_[6]  = TMath::Sign((float)+1., tau->dxy());
+      mvaInput_[7]  = std::sqrt(std::fabs(std::min((float)1., tau->dxy())));
```

Comment:
agreed -- I'll check on the RelVals whether something like tau->dxy()<0 can happen.